packages/runtime/src/components/hooks/useMediaQuery.ts

```diff
@@ -1,32 +1,47 @@
 import { useSyncExternalStore } from 'use-sync-external-store/shim'
 import { DeviceOverride } from '../../prop-controllers'
-import { DEVICES, findDeviceOverride, getDeviceMediaQuery } from '../utils/devices'
+import {
+  Breakpoints,
+  DEFAULT_BREAKPOINTS,
+  findBreakpointOverride,
+  getDeviceMediaQuery,
+} from '../../state/modules/breakpoints'
+import { useBreakpoints } from '../../runtimes/react'
+import { useCallback } from 'react'
 
-const DEVICE_QUERIES = DEVICES.map(device => ({
-  id: device.id,
-  query: getDeviceMediaQuery(device).replace('@media', ''),
-}))
-
-function subscribe(notify: () => void) {
-  const cleanUpFns = DEVICE_QUERIES.map(q => {
-    const mediaQueryList = window.matchMedia(q.query)
-    mediaQueryList.addEventListener('change', notify)
-    return () => mediaQueryList.removeEventListener('change', notify)
-  })
-  return () => cleanUpFns.forEach(fn => fn())
-}
+const getDeviceQueries = (breakpoints: Breakpoints) =>
+  breakpoints.map(device => ({
+    id: device.id,
+    query: getDeviceMediaQuery(device).replace('@media', ''),
+  }))
 
 export function useMediaQuery<S>(responsiveValue?: Array<DeviceOverride<S>>): S | undefined {
-  const getServerSnapshot = () => findDeviceOverride(responsiveValue, DEVICES[0].id)?.value
+  const breakpoints = useBreakpoints()
+  const subscribe = useCallback(
+    (notify: () => void) => {
+      const cleanUpFns = getDeviceQueries(breakpoints).map(q => {
+        const mediaQueryList = window.matchMedia(q.query)
+        mediaQueryList.addEventListener('change', notify)
+        return () => mediaQueryList.removeEventListener('change', notify)
+      })
+      return () => cleanUpFns.forEach(fn => fn())
+    },
+    [breakpoints],
+  )
+  const getServerSnapshot = () =>
+    findBreakpointOverride(breakpoints, responsiveValue, DEFAULT_BREAKPOINTS[0].id)?.value
 
   function getSnapshot() {
-    const deviceId: string = DEVICE_QUERIES.reduce((matchedDevice, deviceQueries) => {
-      if (window.matchMedia(deviceQueries.query).matches) {
-        return deviceQueries.id
-      }
-      return matchedDevice
-    }, DEVICE_QUERIES[0].id)
-    return findDeviceOverride(responsiveValue, deviceId)?.value
+    const deviceId: string = getDeviceQueries(breakpoints).reduce(
+      (matchedDevice, deviceQueries) => {
+        if (window.matchMedia(deviceQueries.query).matches) {
+          return deviceQueries.id
+        }
+        return matchedDevice
+      },
+      DEFAULT_BREAKPOINTS[0].id,
```

Comment:
I think it's ok here not to assume an initial value anymore if the breakpoints are dynamic. In the edge case that there are no breakpoints, then the fallback strategy should take care of it.

Something we need to discuss is how we'll handle that. In the past i've thought of defining a `null` or "special" breakpoint ID which is the "fallback", but perhaps we don't need to use `null`. Here's an idea:

First, we never allow there to be 0 breakpoints, if 0 breakpoints are registered we throw an error, we need at least 1. Furthermore, we store in the state a "base" breakpoint. We can compute that automatically by sorting by width if we're doing a desktop-first design strategy but we could also have other strategies, like mobile-first design. We'll either keep that "base" endpoint in the state or if we're computing it, it can just be a selector.

Then, whenever we're wanting to get the "base" style, or what we call "desktop" right now, we can get the "base" style from state. In this way, we don't need to hardcode `DEFAULT_BREAKPOINTS[0].id` (i.e., `'desktop'`) and things can be dynamic based on the breakpoints